sql/core/src/main/scala/org/apache/spark/sql/execution/EvalNthMatchPocExec.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution
+
+import scala.collection.mutable
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression, Literal, NthMatch, SafeProjection}
+import org.apache.spark.sql.catalyst.trees.TreePattern.NTH_MATCH
+import org.apache.spark.unsafe.types.UTF8String
+
+case class EvalNthMatchPocExec(
+    predicate: Expression,
+    child: SparkPlan)
+  extends UnaryExecNode {
+
+  private val matchedRows: mutable.ArrayBuffer[InternalRow] = mutable.ArrayBuffer[InternalRow]()
+
+  // FIXME: added testing data
+  private def addTestData(employeeName: String, organization: String): Unit = {
+    matchedRows.append(
+      InternalRow(
+        UTF8String.fromString(employeeName),
+        UTF8String.fromString(organization)
+      )
+    )
+  }
+
+  addTestData("Hello", "lakestore")
+  addTestData("World", "streaming")
+  addTestData("Joe", "pyspark")
+  addTestData("Hello", "sql")
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    child.execute().mapPartitions { iter =>
+      logWarning(s"DEBUG: predicate: $predicate")
+
+      val predicateResolved = predicate
+        .transformUpWithPruning(_.containsAnyPattern(NTH_MATCH)) {
+
+        case NthMatch(input, offset) =>
+          val proj = SafeProjection.create(Seq(input), child.output)
```

Comment:
The creation of projection can be done among partitions, as long as we cache it and refer to it later.